### PR TITLE
Improve contest ui

### PR DESCRIPTION
**Pull request by breeswish**

### CODE REVIEW

File: vj4/ui/templates/discussion_main_or_node.html [hunk 1]
```diff
@@ -40,7 +39,9 @@ <h1 class="discussion__title"><a href="{{ reverse_url('discussion_detail', did=d
   </div>
   <div class="medium-3 columns">
   {% if vnode['doc_type'] == vj4.model.document.TYPE_PROBLEM %}
-    {{ problem.render_sidebar(vnode, udoc) }}{# TODO: where is udoc? #}
+    {% with pdoc=vnode, owner_udoc=udoc %}{# TODO: where is udoc? #}
```
Comment (breeswish) [inline vj4/ui/templates/discussion_main_or_node.html]: @twd2

File: vj4/ui/templates/partials/contest_sidebar.html [hunk 2]
```diff
@@ -0,0 +1,67 @@
+{% import "components/user.html" as user with context %}
+{% import "components/contest.html" as contest with context %}
+<div class="section side">
+{% if page_name != 'contest_detail' %}
+  <a class="contest-sidebar__bg" href="{{ reverse_url('contest_detail', tid=tdoc['_id']) }}">
+    <div class="section__header">
+      <h1>{{ tdoc['title'] }}</h1>
+    </div>
+    <div class="section__body">
+      <!-- TODO -->
```
Comment (breeswish) [inline vj4/ui/templates/partials/contest_sidebar.html]: @twd2

File: vj4/ui/templates/partials/contest_sidebar.html [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    {% if page_name != 'contest_detail' %}
+      <li class="menu__item"><a class="menu__link" href="{{ reverse_url('contest_detail', tid=tdoc['_id']) }}">
+        <span class="icon icon-award"></span> {{ _('View Contest') }}
+      </a></li>
+    {% endif %}
+    {% if handler.has_perm(vj4.model.builtin.PERM_ATTEND_CONTEST) %}
+    {% if not attended %}
+      <li class="menu__item">
+        <form method="POST">
+          <!-- TODO -->
```
Comment (breeswish) [inline vj4/ui/templates/partials/contest_sidebar.html]: @twd2 attend button can be in every page now. submit to self is not approachable.

File: vj4/ui/templates/contest_main.html [hunk 4]
```diff
@@ -50,40 +51,83 @@ <h2 class="status_title">{{ _(title) }}</h2>
...
+    {% else %}
       <div class="section__header">
         <h1 class="section__title">{{ _('All Contests') }}</h1>
+        <div class="section__tools">
+          {{ _('Show') }}:
+          <select class="inline select">
+            <option>All</option>
+            <option>Only OI</option>
+            <option>Only ACM/ICPC</option>
+            <!-- TODO: dynamic list -->
```
Comment (breeswish) [inline vj4/ui/templates/contest_main.html]: @twd2 We need to generate the list dynamically. Of course we should also implement contest filter.

File: vj4/ui/templates/contest_main.html [hunk 5]
```diff
@@ -50,40 +51,83 @@ <h2 class="status_title">{{ _(title) }}</h2>
...
-                  </ul>
-                </div>
+      <ol class="contest__list">
+      {% for tdoc in tdocs %}
+        <li class="contest__item contest-type--{{ vj4.constant.contest.RULE_ID[tdoc['rule']] }}">
+          <div class="media">
+            <div class="media__left medium">
+              <div class="contest__date numbox">
+                <div class="numbox__num large">4</div>
+                <div class="numbox__text">2016.06</div>
```
Comment (breeswish) [inline vj4/ui/templates/contest_main.html]: @twd2 TODO: this should be a formatted contest start date

File: vj4/ui/templates/contest_main.html [hunk 6]
```diff
@@ -50,40 +51,83 @@ <h2 class="status_title">{{ _(title) }}</h2>
...
               </div>
-            </li></a>
-          {% endif %}
+            </div>
+            <div class="media__body medium">
+              <h1 class="contest__title"><a href="{{ reverse_url('contest_detail', tid=tdoc['doc_id']) }}" data-emoji-enabled>{{ tdoc['title'] }}</a></h1>
+              <ul class="contest__supplementary">
+                <li>
+                  <a href="#" class="contest-type-tag"><span class="icon icon-award"></span>{{ vj4.constant.contest.RULE_TEXTS[tdoc['rule']] }}</a>
```
Comment (breeswish) [inline vj4/ui/templates/contest_main.html]: @twd2 When clicking links, a filter should be applied

File: vj4/ui/templates/partials/contest_sidebar.html (hunk 2 above)
Comment (twd2) [inline vj4/ui/templates/partials/contest_sidebar.html]: done

File: vj4/ui/templates/partials/contest_sidebar.html (hunk 3 above)
Comment (twd2) [inline vj4/ui/templates/partials/contest_sidebar.html]: done

File: vj4/ui/templates/contest_main.html (hunk 4 above)
Comment (twd2) [inline vj4/ui/templates/contest_main.html]: done1

File: vj4/ui/templates/contest_main.html (hunk 5 above)
Comment (twd2) [inline vj4/ui/templates/contest_main.html]: done

File: vj4/ui/templates/contest_main.html (hunk 6 above)
Comment (twd2) [inline vj4/ui/templates/contest_main.html]: done

File: vj4/ui/templates/contest_main.html (hunk 4 above)
Comment (twd2) [inline vj4/ui/templates/contest_main.html]: done2

File: vj4/ui/templates/contest_main.html [hunk 7]
```diff
@@ -42,48 +47,77 @@ <h2 class="status_title">{{ _(title) }}</h2>
...
-    {% elif handler.is_ready(tdoc) %}
+  {% elif handler.is_ready(tdoc) %}
     {{ render_important_contest('ready', 'Ready (☆▽☆)', tdoc) }}
-    {% endif %}
-    {% endfor %}
+  {% endif %}
+  {% endfor %}
     <div class="section">
       <div class="section__header">
         <h1 class="section__title">{{ _('All Contests') }}</h1>
```
Comment (breeswish) [inline vj4/ui/templates/contest_main.html]: @twd2 Should show `All OI Contests` when filtering for OI, etc.

Comment (twd2) [inline vj4/ui/templates/contest_main.html]: 好

**Pull request merged**